Fixes race condition; improves error reporting

rickb777 opened the pull request.

Commit:
Fix an observed race condition where the server code and test code would fail to see sequential operation of the shared boolean (presumably due to CPU cache staleness etc)

Commit:
StubMethod constructor is now public, allowing other methods to be constructed as needed. This now supports WebDav, for example.

Commit:
Enhanced the message reported when no expectation has matched so that it is easier to diagnose what error caused the non-matching outcome.

tarttelin commented on src/main/java/com/pyruby/stubserver/StubServer.java:
```diff
@@ -213,8 +217,21 @@ public void handle(String target, HttpServletRequest httpServletRequest, HttpSer
+                Enumeration headerValues = httpServletRequest.getHeaders(headerName);
+                message.append(header(headerName, headerValues)).append("\n");
             }
+            Expectation.CannedResponse nullResponse = new Expectation.CannedResponse(404, "text/plain", message.toString(), null);
```
This will break a number of existing users, as it changes the default from a 200 to a 404. The mockito model follows the 'Return null for unmatched' which is where this library drew it's inspiration from.

I would favour a switch or config that specified a 'Fail on unmatched' as an alternative.

Commit:
Altered the status-if-unmatched so that it defaults to 200 (the original behaivour). It is now possible to set a different code to be used when expectations are unmatched, such asa 404.

rickb777 commented:
Hi Chris,

I've taken your requirement about the default 200 code on board and added an option to sat any required value (404 being the obvious choice).

I've also fixed a deficiency in the query string handling. But note that there is no reporting of PUT/POST body data in this implementation. I don't need this at the moment, but it might be worth revisiting later.

Rick
